jaxrs/src/main/java/com/facebook/airlift/jaxrs/thrift/ThriftMapper.java

```diff
@@ -0,0 +1,157 @@
+     * blacklisting things we are not to handle.
+     */
+    private static final ImmutableSet<Class<?>> IO_CLASSES = ImmutableSet.<Class<?>>builder()
+            .add(InputStream.class)
+            .add(java.io.Reader.class)
+            .add(OutputStream.class)
+            .add(java.io.Writer.class)
+            .add(byte[].class)
+            .add(char[].class)
+            .add(javax.ws.rs.core.StreamingOutput.class)
+            .add(Response.class)
+            .build();
+
+    public static final Logger log = Logger.get(ThriftMapper.class);
+
+    private final ThriftCodecManager thriftCodecManager;
+
+    @Inject
+    public ThriftMapper(ThriftCodecManager thriftCodecManager)
+    {
+        this.thriftCodecManager = thriftCodecManager;
+    }
+
+    @Override
+    public boolean isReadable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType)
+    {
+        return canReadOrWrite(type);
+    }
+
+    @Override
+    public boolean isWriteable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType)
+    {
+        return canReadOrWrite(type);
+    }
+
+    private static boolean canReadOrWrite(Class<?> type)
+    {
+        if (IO_CLASSES.contains(type)) {
+            return false;
+        }
+        for (Class<?> ioClass : IO_CLASSES) {
+            if (ioClass.isAssignableFrom(type)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    @Override
+    public Object readFrom(Class<Object> type,
+            Type genericType,
+            Annotation[] annotations,
+            MediaType mediaType,
+            MultivaluedMap<String, String> httpHeaders,
+            InputStream inputStream)
+            throws IOException
+    {
+        try {
+            ThriftCodec<?> codec = thriftCodecManager.getCodec(type);
+            TProtocol tProtocol = new TFacebookCompactProtocol(new TInputStreamTransport(inputStream));
+            return codec.read(tProtocol);
+        }
+        catch (Exception e) {
+            // we want to return a 400 for bad Thrift but not for a real IO exception
+            if (e instanceof IOException && !(e instanceof EOFException)) {
+                throw (IOException) e;
+            }
+            // log the exception at debug so it can be viewed during development
+            // Note: we are not logging at a higher level because this could cause a denial of service
+            log.debug(e, "Invalid Thrift for Java type %s", type);
+            // Invalid thrift request. Throwing exception so the response code can be overridden using a mapper.
+            throw new ThriftMapperParsingException(type, e);
+        }
+    }
+
+    @Override
+    public long getSize(Object value, Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType)
+    {
+        // In general figuring output size requires actual writing; usually not
+        // worth it to write everything twice.
```

Comment:
`getSize` seems to be deprecated from JAX-RS 2.0. You can maybe remove this comment since returning -1 is what is expected. 

from [javadocs](https://docs.oracle.com/javaee/7/api/javax/ws/rs/ext/MessageBodyWriter.html#getSize-T-java.lang.Class-java.lang.reflect.Type-java.lang.annotation.Annotation:A-javax.ws.rs.core.MediaType-)

> As of JAX-RS 2.0, the method has been deprecated and the value returned by the method is ignored by a JAX-RS runtime. All MessageBodyWriter implementations are advised to return -1 from the method. Responsibility to compute the actual Content-Length header value has been delegated to JAX-RS runtime.